analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs

```diff
@@ -0,0 +1,152 @@
+{{
+    value.ToString(); // Add another constraint to 'value'
+    value--;
+}}
+Tag(""End"", value);";
+            var validator = SETestContext.CreateCS(code, ", int[] items", new AddConstraintOnInvocationCheck()).Validator;
+            validator.ValidateExitReachCount(2);
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
+                .And.ContainSingle(x => x != null && x.HasConstraint(TestConstraint.First) && !x.HasConstraint(BoolConstraint.True));
+        }
+
+        [TestMethod]
+        public void Loops_InstructionVisitedMaxTwice_ForEachOriginalState()
+        {
+            const string code = @"
+var value = 42;
+bool condition;
+if(boolParameter)      // This generates two different ProgramStates, each tracks its own visits
+    condition = true;
+else
+    condition = false;
+do
+{
+    value.ToString(); // Add another constraint to 'value'
+} while (true);
+Tag(""End"", value);";
+            var validator = SETestContext.CreateCS(code, ", int[] items", new BoolTestCheck(), new AddConstraintOnInvocationCheck()).Validator;
+            validator.ValidateExitReachCount(4);
+            var states = validator.TagStates("End");
+            var condition = states.SelectMany(x => x.SymbolsWith(BoolConstraint.False)).First();    // "False" is never set for "value"
+            var value = states.SelectMany(x => x.SymbolsWith(TestConstraint.First)).First();        // "First" is never set for "condition"
+            states.Should().HaveCount(4)
+                .And.ContainSingle(x => x[condition].HasConstraint(BoolConstraint.True) && x[value].HasConstraint(TestConstraint.First) && !x[value].HasConstraint(BoolConstraint.True))
+                .And.ContainSingle(x => x[condition].HasConstraint(BoolConstraint.True) && x[value].HasConstraint(TestConstraint.First) && x[value].HasConstraint(BoolConstraint.True))
+                .And.ContainSingle(x => x[condition].HasConstraint(BoolConstraint.False) && x[value].HasConstraint(TestConstraint.First) && !x[value].HasConstraint(BoolConstraint.True))
+                .And.ContainSingle(x => x[condition].HasConstraint(BoolConstraint.False) && x[value].HasConstraint(TestConstraint.First) && x[value].HasConstraint(BoolConstraint.True));
+        }
+
+        [DataTestMethod]
+        [DataRow("value > 0")]
+        [DataRow("true")]
+        public void DoLoop_InstructionVisitedMaxTwice(string condition)
+        {
+            var code = $@"
+var value = 42;
+do
+{{
+    value.ToString(); // Add another constraint to 'value'
+    value--;
+}} while ({condition});
+Tag(""End"", value);";
+            var validator = SETestContext.CreateCS(code, new AddConstraintOnInvocationCheck()).Validator;
+            validator.ValidateExitReachCount(2);
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x.HasConstraint(TestConstraint.First) && !x.HasConstraint(BoolConstraint.True));
```

Comment:
Did you intentionally tested only one of the values?

Reply:
No. But I intentionally asserted the count first to make it easier to spot :)